File: algorithm/binary search/BOJ 1920.py
#1920 수 찾기
n = int(input())
a = list(map(int, input().split())) #반복문 돌 놈
m = int(input())
arr = list(map(int, input().split())) # 있는지 확인할 놈
a.sort()


for i in arr:
    left =0
    right = n-1
    while True:
        mid = (left+right)//2
        if left > right:
            print(0)
            break
        if a[mid] == i:
            print(1)
            break
        elif i> a[mid]:
            left = mid +1
        elif i<a[mid]:
            right = mid - 1
    

File: algorithm/binary search/boj1920.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class boj1920 {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		
		int N = Integer.parseInt(br.readLine());
		int[] arr = new int[N];

		StringTokenizer st = new StringTokenizer(br.readLine());
		for (int i = 0; i < N; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		Arrays.sort(arr);

		int M = Integer.parseInt(br.readLine());
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < M; i++) {
			int num = Integer.parseInt(st.nextToken());

			int left = 0;
			int right = N-1;
			boolean result = false;
			while (left<=right) {
				int mid = (left + right)/2;
				
				if (num==arr[mid]) {
					result = true;
					break;
				} else if (num<arr[mid]) {
					right = mid-1;
				} else if (num>arr[mid]) {
					left = mid+1;
				}		
			}
			
			if (result) sb.append(1).append("\n");
			else sb.append(0).append("\n");			
		}
		System.out.println(sb);
	}
}
